src/main/java/me/kavin/piped/utils/ResponseHelper.java

```diff
@@ -1158,6 +1215,39 @@ public static void handleNewVideo(String url, long time, me.kavin.piped.utils.ob
         }
     }
 
+    private static void pruneUnusedPlaylistVideos() {
+        try (Session s = DatabaseSessionFactory.createSession()) {
+            CriteriaQuery<me.kavin.piped.utils.obj.db.Playlist> plQuery =
+                s.getCriteriaBuilder().createQuery(me.kavin.piped.utils.obj.db.Playlist.class);
+            plQuery.select(plQuery.from(me.kavin.piped.utils.obj.db.Playlist.class));
+            List<me.kavin.piped.utils.obj.db.Playlist> playlists = s.createQuery(plQuery).getResultList();
+
+            CriteriaQuery<PlaylistVideo> pvQuery = s.getCriteriaBuilder().createQuery(PlaylistVideo.class);
+            pvQuery.select(pvQuery.from(PlaylistVideo.class));
+            List<PlaylistVideo> playlistVideos = s.createQuery(pvQuery).getResultList();
+
+            Iterator<PlaylistVideo> pvIter = playlistVideos.iterator();
+
+            outer:
+            while (pvIter.hasNext()) {
+                PlaylistVideo pv = pvIter.next();
+
+                for (me.kavin.piped.utils.obj.db.Playlist pl : playlists) {
+                    for (PlaylistVideo v : pl.getVideos()) {
+                        if (v.getId().equals(pv.getId())) {
+                            continue outer;
+                        }
+                    }
+                }
+
+                s.delete(pv);
+            }

```

Comment:
I feel this can be optimized to a single query by using a query to delete videos by using `not in` / isMember in a `CriteriaBuilder` with a subquery. This could use a lot of session time incase the database has a lot of videos/playlists.